python/nav/web/auth.py

```diff
@@ -24,57 +24,106 @@
 logger = logging.getLogger("nav.web.auth")
 
 
-def authenticate(username, password):
-    '''Authenticate username and password against database.
+def authenticate_account(username=None, password=None):
+    """
+    Authenticate username and password against database
+
     Returns account object if user was authenticated, else None.
-    '''
-    # FIXME Log stuff?
-    auth = False
-    account = None
+    """
+    if not username and password:
+        return None
 
-    # Try to find the account in the database. If it's not found we can try
-    # LDAP.
     try:
         account = Account.objects.get(login__iexact=username)
     except Account.DoesNotExist:
-        if ldapauth.available:
-            user = ldapauth.authenticate(username, password)
-            # If we authenticated, store the user in database.
-            if user:
-                account = Account(
-                    login=user.username,
-                    name=user.get_real_name(),
-                    ext_sync='ldap'
-                )
-                account.set_password(password)
-                account.save()
-                # We're authenticated now
-                auth = True
-
-    if account and account.locked:
-        logger.info("Locked user %s tried to log in", account.login)
+        return None
+
+    if account.check_password(password):
+        return account
+
+    # Password was incorrect
+    return None
+
+
+def authenticate_ldap(username=None, password=None):
+    """
+    Authenticate username and password against LDAP, if available
+
+    Returns account object if user was authenticated, else None.
+    """
+    if not username and password:
+        return None
+
+    if not ldapauth.available:
+        return None
+
+    try:
+        ldapuser = ldapauth.authenticate(username, password)
+    except ldapauth.NoAnswerError:
+        # LDAP unreachable, fallback
+        return None
+
+    if ldapuser is False:
+        # This user does not exist in LDAP, fallback
+        return None
```

Comment:
Here you do not distinguish between "LDAP does not answer" in which case we should fall back, and "LDAP says the user does not exist" in which case we should fail the log in. Otherwise users would retain their NAV access even when deleted from LDAP. We don't want that.

Reply:
If the ldap-user does not exist, should the Account then also be set to be locked, maybe?

Reply:
> If the ldap-user does not exist, should the Account then also be set to be locked, maybe?

It did not previously, but I think that would be a *useful improvement* over the current behavior.

The current behavior means that if the LDAP server goes down, deleted users can log in to NAV using their last known good password.

Reply:
That would actually work with the new logic, the `authenticate_account` would then toss the login-attempt, and the user may not log in if the ldap-server is unavailable.

Another way would be to separate between None and False, None for "fallback to next method" and False for "not authenticated, abort login-attempt".

Reply:
I support locking the account in this case, but it should not be introduced as a semantic change as part of a refactor commit. If any piece of code should have proper traceability (and test coverage) it should be the authentication code

Reply:
Both then, probably. Returning False (or raising an exception) to abort, then locking in a later patch.

Reply:
False for failed auth is pushed.